doc/diagnostics/VMState.md

```diff
@@ -0,0 +1,88 @@
+<!--
+Copyright (c) 2018, 2018 IBM Corp. and others
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+**VM state**
+
+The VM state is a numeric field stored in J9VMThread->omrVMThread->vmState that uses 32-bits to encode the current state of the thread. See the `J9VMSTATE_` defined constants in the code. For example, if the thread is currently JIT compiling a method, the VM state indicates which compilation phase the compiler is in. If the VM crashes for any reason, the VM state is reported via standard error and in the javacore dump (in which it is referred to as "VM flags"), and serves as a first-order indicator of the failing component during problem determination. 
```

Comment:
> See the `J9VMSTATE_` defined constants in the code.
Is it possible to link to the file directly in the Markdown in which these are defined?

Reply:
I could, but it might become out of date as the code could move without these docs being updated.

Reply:
I second @fjeremic request - a direct link will make it very easy to see the constants and we can always update the link later if we find it's been broken.

Reply:
Added a direct link to j9nonbuilder.h